File: app/api/repository_intelligence.py
"""Workspace-scoped Repository Intelligence read APIs (RI-007)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, ConfigDict, Field

from app.api.workspace_auth import WorkspaceAccess, require_workspace_access
from app.db.base import AsyncSessionLocal
from app.services.repository_intelligence.read_service import (
    DEFAULT_HISTORY_LIMIT,
    DEFAULT_PORTFOLIO_LIMIT,
    MAX_GRAPH_EDGES,
    MAX_HISTORY_LIMIT,
    MAX_PORTFOLIO_LIMIT,
    build_repository_intelligence_detail,
    build_repository_intelligence_graph,
    build_repository_intelligence_history,
    build_repository_intelligence_portfolio,
)


router = APIRouter(
    prefix="/api/v1/workspaces/{workspace_id}/repository-intelligence",
    tags=["repository-intelligence"],
)


class _ReadModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RepositoryIntelligenceCapabilitiesRead(_ReadModel):
    provider_calls: Literal[False]
    repository_reads: Literal[False]
    target_execution: Literal[False]
    external_writes: Literal[False]
    llm_used: Literal[False]
    human_resolution_writes: Literal[False]


class RepositoryIdentityRead(_ReadModel):
    id: UUID
    provider: Literal["github"]
    external_id: str
    name: str
    full_name: str
    default_branch: str | None = None
    visibility: str | None = None
    archived: bool
    source_url: str | None = None
    last_activity_at: datetime | None = None


class RepositoryLatestAuditRead(_ReadModel):
    id: UUID
    audit_level: Literal["L0", "L1", "L2"]
    target_status: Literal["exact", "unavailable"]
    commit_sha: str | None = None
    metadata_snapshot_id: str | None = None
    profile: str
    engine_version: str
    status: Literal["succeeded", "partial", "failed", "cancelled"]
    coverage_status: Literal["complete", "partial"]
    reconciliation_applied: bool
    completed_at: datetime
    artifact_status: Literal["retained", "purged"]


class RepositoryOpenFindingsRead(_ReadModel):
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    info: int = 0


class RepositoryPortfolioItemRead(RepositoryIdentityRead):
    purpose_summary: str | None = None
    operational_summary: str | None = None
    repository_type: str
    purpose_status: str
    purpose_confidence: float
    product_candidates: list[str] = Field(default_factory=list)
    owner_candidates: list[str] = Field(default_factory=list)
    has_confirmed_owner: bool
    latest_audit: RepositoryLatestAuditRead | None = None
    open_findings: RepositoryOpenFindingsRead
    open_findings_total: int
    outbound_relationship_count: int
    inbound_relationship_count: int
    unknown_count: int
    pending_confirmation_count: int
    has_stale_intelligence: bool


class RepositoryPortfolioSummaryRead(_ReadModel):
    repositories: int
    analyzed_repositories: int
    repositories_with_open_findings: int
    repositories_with_stale_intelligence: int
    current_relationships: int
    blocking_unknowns: int
    pending_confirmations: int


class RepositoryPortfolioLimitsRead(_ReadModel):
    repositories: int


class RepositoryPortfolioResponse(_ReadModel):
    workspace_id: UUID
    mode: Literal["repository_intelligence_read_only"] = (
        "repository_intelligence_read_only"
    )
    source: Literal["ri_006_persistence"] = "ri_006_persistence"
    summary: RepositoryPortfolioSummaryRead
    repositories: list[RepositoryPortfolioItemRead] = Field(default_factory=list)
    limits: RepositoryPortfolioLimitsRead
    truncated: bool
    capabilities: RepositoryIntelligenceCapabilitiesRead
    warnings: list[str] = Field(default_factory=list)


class RepositoryEvidenceRead(_ReadModel):
    id: UUID
    role: Literal["supporting", "contradicting"]
    kind: str
    source: str
    ref: str | None = None
    record_id: UUID
    url: str | None = None
    confidence: float


class RepositoryFactRead(_ReadModel):
    id: UUID
    fact_type: str
    claim_id: str
    value: dict[str, Any]
    claim_status: str
    confidence: float
    lifecycle_status: Literal["current", "stale"]
    human_resolution_status: Literal["pending", "confirmed", "rejected"]
    first_seen_at: datetime | None = None
    last_seen_at: datetime | None = None
    stale_at: datetime | None = None
    evidence: list[RepositoryEvidenceRead] = Field(default_factory=list)


class RepositoryGraphReferenceRead(_ReadModel):
    id: UUID
    full_name: str


class RepositoryRelationshipRead(_ReadModel):
    id: UUID
    direction: Literal["outbound", "inbound"]
    from_repository: RepositoryGraphReferenceRead
    to_repository: RepositoryGraphReferenceRead | None = None
    target_full_name: str
    relationship_type: str
    resolution_status: Literal["canonical", "candidate"]
    summary: str | None = None
    claim_status: Literal["observed", "inferred"]
    confidence: float
    lifecycle_status: Literal["current", "stale"]
    human_resolution_status: Literal["pending", "confirmed", "rejected"]
    first_seen_at: datetime | None = None
    last_seen_at: datetime | None = None
    stale_at: datetime | None = None
    evidence: list[RepositoryEvidenceRead] = Field(default_factory=list)


class RepositoryFindingRead(_ReadModel):
    id: UUID
    finding_id: str
    rule_id: str
    category: str
    severity: Literal["info", "low", "medium", "high", "critical"]
    confidence: float
    status: str
    title: str
    summary: str
    recommended_next_step: str | None = None
    first_seen_at: datetime | None = None
    last_seen_at: datetime | None = None
    resolved_at: datetime | None = None
    evidence: list[RepositoryEvidenceRead] = Field(default_factory=list)


class RepositoryContradictionFactRead(_ReadModel):
    id: UUID
    fact_type: str
    claim_id: str
    value: dict[str, Any]
    claim_status: str


class RepositoryContradictionRead(_ReadModel):
    id: UUID
    contradiction_id: str
    status: Literal["current", "resolved"]
    confidence: float
    summary: str
    left_fact: RepositoryContradictionFactRead | None = None
    right_fact: RepositoryContradictionFactRead | None = None
    first_seen_at: datetime | None = None
    last_seen_at: datetime | None = None
    resolved_at: datetime | None = None
    evidence: list[RepositoryEvidenceRead] = Field(default_factory=list)


class RepositoryConfirmationRead(_ReadModel):
    kind: Literal["fact", "relationship"]
    id: UUID
    label: str
    claim_status: str
    human_resolution_status: Literal["pending", "confirmed", "rejected"]
    evidence: list[RepositoryEvidenceRead] = Field(default_factory=list)


class RepositoryDetailTruncatedRead(_ReadModel):
    facts: bool
    relationships: bool
    findings: bool
    contradictions: bool
    confirmation_queue: bool


class RepositoryDetailResponse(_ReadModel):
    workspace_id: UUID
    mode: Literal["repository_intelligence_read_only"] = (
        "repository_intelligence_read_only"
    )
    source: Literal["ri_006_persistence"] = "ri_006_persistence"
    repository: RepositoryIdentityRead
    purpose: RepositoryFactRead | None = None
    latest_audit: RepositoryLatestAuditRead | None = None
    facts: list[RepositoryFactRead] = Field(default_factory=list)
    relationships: list[RepositoryRelationshipRead] = Field(default_factory=list)
    findings: list[RepositoryFindingRead] = Field(default_factory=list)
    contradictions: list[RepositoryContradictionRead] = Field(default_factory=list)
    unknowns: list[RepositoryFactRead] = Field(default_factory=list)
    confirmation_queue: list[RepositoryConfirmationRead] = Field(
        default_factory=list
    )
    limitations: list[str] = Field(default_factory=list)
    truncated: RepositoryDetailTruncatedRead
    capabilities: RepositoryIntelligenceCapabilitiesRead


class RepositoryHistoryRunRead(_ReadModel):
    id: UUID
    audit_level: Literal["L0", "L1", "L2"]
    target_status: Literal["exact", "unavailable"]
    commit_sha: str | None = None
    metadata_snapshot_id: str | None = None
    profile: str
    policy_hash: str
    engine_version: str
    status: Literal["succeeded", "partial", "failed", "cancelled"]
    coverage_status: Literal["complete", "partial"]
    completed_checks: list[str] = Field(default_factory=list)
    failed_checks: list[str] = Field(default_factory=list)
    skipped_checks: list[str] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)
    reconciliation_applied: bool
    artifact_count: int
    artifact_status: Literal["retained", "purged"]
    started_at: datetime
    completed_at: datetime


class RepositoryHistoryResponse(_ReadModel):
    workspace_id: UUID
    mode: Literal["repository_intelligence_read_only"] = (
        "repository_intelligence_read_only"
    )
    source: Literal["ri_006_persistence"] = "ri_006_persistence"
    repository: RepositoryIdentityRead
    runs: list[RepositoryHistoryRunRead] = Field(default_factory=list)
    limit: int
    truncated: bool
    capabilities: RepositoryIntelligenceCapabilitiesRead


class RepositoryGraphNodeRead(_ReadModel):
    id: UUID
    full_name: str
    repository_type: str
    archived: bool
    open_findings_total: int
    has_stale_intelligence: bool
    latest_audit_at: datetime | None = None


class RepositoryGraphEdgeRead(_ReadModel):
    id: UUID
    from_repository_id: UUID
    from_repository_full_name: str
    to_repository_id: UUID | None = None
    target_full_name: str
    relationship_type: str
    resolution_status: Literal["canonical", "candidate"]
    claim_status: Literal["observed", "inferred"]
    human_resolution_status: Literal["pending", "confirmed", "rejected"]
    confidence: float
    summary: str | None = None


class RepositoryGraphSummaryRead(_ReadModel):
    nodes: int
    edges: int
    observed_edges: int
    inferred_edges: int
    candidate_edges: int


class RepositoryGraphTruncatedRead(_ReadModel):
    nodes: bool
    edges: bool


class RepositoryGraphResponse(_ReadModel):
    workspace_id: UUID
    mode: Literal["repository_intelligence_read_only"] = (
        "repository_intelligence_read_only"
    )
    source: Literal["ri_006_persistence"] = "ri_006_persistence"
    nodes: list[RepositoryGraphNodeRead] = Field(default_factory=list)
    edges: list[RepositoryGraphEdgeRead] = Field(default_factory=list)
    summary: RepositoryGraphSummaryRead
    truncated: RepositoryGraphTruncatedRead
    capabilities: RepositoryIntelligenceCapabilitiesRead


@router.get("", response_model=RepositoryPortfolioResponse)
async def get_repository_intelligence_portfolio(
    workspace_id: UUID,
    limit: int = Query(
        default=DEFAULT_PORTFOLIO_LIMIT,
        ge=1,
        le=MAX_PORTFOLIO_LIMIT,
    ),
    access: WorkspaceAccess = Depends(require_workspace_access),
) -> RepositoryPortfolioResponse:
    """Read the bounded Repository Intelligence portfolio for one workspace."""

    _ = access
    async with AsyncSessionLocal() as session:
        payload = await build_repository_intelligence_portfolio(
            session=session,
            workspace_id=workspace_id,
            limit=limit,
        )
    return RepositoryPortfolioResponse.model_validate(payload)


@router.get("/graph", response_model=RepositoryGraphResponse)
async def get_repository_intelligence_graph(
    workspace_id: UUID,
    repository_limit: int = Query(
        default=MAX_PORTFOLIO_LIMIT,
        ge=1,
        le=MAX_PORTFOLIO_LIMIT,
    ),
    edge_limit: int = Query(
        default=MAX_GRAPH_EDGES,
        ge=1,
        le=MAX_GRAPH_EDGES,
    ),
    access: WorkspaceAccess = Depends(require_workspace_access),
) -> RepositoryGraphResponse:
    """Read the bounded current directional repository graph."""

    _ = access
    async with AsyncSessionLocal() as session:
        payload = await build_repository_intelligence_graph(
            session=session,
            workspace_id=workspace_id,
            repository_limit=repository_limit,
            edge_limit=edge_limit,
        )
    return RepositoryGraphResponse.model_validate(payload)


@router.get(
    "/repositories/{repository_id}",
    response_model=RepositoryDetailResponse,
)
async def get_repository_intelligence_detail(
    workspace_id: UUID,
    repository_id: UUID,
    access: WorkspaceAccess = Depends(require_workspace_access),
) -> RepositoryDetailResponse:
    """Read facts, edges, findings, unknowns, and sanitized evidence."""

    _ = access
    async with AsyncSessionLocal() as session:
        payload = await build_repository_intelligence_detail(
            session=session,
            workspace_id=workspace_id,
            repository_id=repository_id,
        )
    if payload is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="repository intelligence not found",
        )
    return RepositoryDetailResponse.model_validate(payload)


@router.get(
    "/repositories/{repository_id}/history",
    response_model=RepositoryHistoryResponse,
)
async def get_repository_intelligence_history(
    workspace_id: UUID,
    repository_id: UUID,
    limit: int = Query(
        default=DEFAULT_HISTORY_LIMIT,
        ge=1,
        le=MAX_HISTORY_LIMIT,
    ),
    access: WorkspaceAccess = Depends(require_workspace_access),
) -> RepositoryHistoryResponse:
    """Read bounded immutable audit history without artifact paths or bodies."""

    _ = access
    async with AsyncSessionLocal() as session:
        payload = await build_repository_intelligence_history(
            session=session,
            workspace_id=workspace_id,
            repository_id=repository_id,
            limit=limit,
        )
    if payload is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="repository intelligence not found",
        )
    return RepositoryHistoryResponse.model_validate(payload)

File: web/lib/repository-intelligence.ts
import { apiFetch } from "./api";
import type { ApiFetchOptions } from "./types";

export type RepositoryIntelligenceCapabilities = {
  provider_calls: false;
  repository_reads: false;
  target_execution: false;
  external_writes: false;
  llm_used: false;
  human_resolution_writes: false;
};

export type RepositoryIntelligenceIdentity = {
  id: string;
  provider: "github";
  external_id: string;
  name: string;
  full_name: string;
  default_branch: string | null;
  visibility: string | null;
  archived: boolean;
  source_url: string | null;
  last_activity_at: string | null;
};

export type RepositoryLatestAudit = {
  id: string;
  audit_level: "L0" | "L1" | "L2";
  target_status: "exact" | "unavailable";
  commit_sha: string | null;
  metadata_snapshot_id: string | null;
  profile: string;
  engine_version: string;
  status: "succeeded" | "partial" | "failed" | "cancelled";
  coverage_status: "complete" | "partial";
  reconciliation_applied: boolean;
  completed_at: string;
  artifact_status: "retained" | "purged";
};

export type RepositoryOpenFindings = {
  critical: number;
  high: number;
  medium: number;
  low: number;
  info: number;
};

export type RepositoryPortfolioItem = RepositoryIntelligenceIdentity & {
  purpose_summary: string | null;
  operational_summary: string | null;
  repository_type: string;
  purpose_status: string;
  purpose_confidence: number;
  product_candidates: string[];
  owner_candidates: string[];
  has_confirmed_owner: boolean;
  latest_audit: RepositoryLatestAudit | null;
  open_findings: RepositoryOpenFindings;
  open_findings_total: number;
  outbound_relationship_count: number;
  inbound_relationship_count: number;
  unknown_count: number;
  pending_confirmation_count: number;
  has_stale_intelligence: boolean;
};

export type RepositoryPortfolioResponse = {
  workspace_id: string;
  mode: "repository_intelligence_read_only";
  source: "ri_006_persistence";
  summary: {
    repositories: number;
    analyzed_repositories: number;
    repositories_with_open_findings: number;
    repositories_with_stale_intelligence: number;
    current_relationships: number;
    blocking_unknowns: number;
    pending_confirmations: number;
  };
  repositories: RepositoryPortfolioItem[];
  limits: {
    repositories: number;
  };
  truncated: boolean;
  capabilities: RepositoryIntelligenceCapabilities;
  warnings: string[];
};

export type RepositoryEvidence = {
  id: string;
  role: "supporting" | "contradicting";
  kind: string;
  source: string;
  ref: string | null;
  record_id: string;
  url: string | null;
  confidence: number;
};

export type RepositoryFact = {
  id: string;
  fact_type: string;
  claim_id: string;
  value: Record<string, unknown>;
  claim_status: string;
  confidence: number;
  lifecycle_status: "current" | "stale";
  human_resolution_status: "pending" | "confirmed" | "rejected";
  first_seen_at: string | null;
  last_seen_at: string | null;
  stale_at: string | null;
  evidence: RepositoryEvidence[];
};

export type RepositoryRelationship = {
  id: string;
  direction: "outbound" | "inbound";
  from_repository: {
    id: string;
    full_name: string;
  };
  to_repository: {
    id: string;
    full_name: string;
  } | null;
  target_full_name: string;
  relationship_type: string;
  resolution_status: "canonical" | "candidate";
  summary: string | null;
  claim_status: "observed" | "inferred";
  confidence: number;
  lifecycle_status: "current" | "stale";
  human_resolution_status: "pending" | "confirmed" | "rejected";
  first_seen_at: string | null;
  last_seen_at: string | null;
  stale_at: string | null;
  evidence: RepositoryEvidence[];
};

export type RepositoryFinding = {
  id: string;
  finding_id: string;
  rule_id: string;
  category: string;
  severity: "info" | "low" | "medium" | "high" | "critical";
  confidence: number;
  status: string;
  title: string;
  summary: string;
  recommended_next_step: string | null;
  first_seen_at: string | null;
  last_seen_at: string | null;
  resolved_at: string | null;
  evidence: RepositoryEvidence[];
};

export type RepositoryContradictionFact = {
  id: string;
  fact_type: string;
  claim_id: string;
  value: Record<string, unknown>;
  claim_status: string;
};

export type RepositoryContradiction = {
  id: string;
  contradiction_id: string;
  status: "current" | "resolved";
  confidence: number;
  summary: string;
  left_fact: RepositoryContradictionFact | null;
  right_fact: RepositoryContradictionFact | null;
  first_seen_at: string | null;
  last_seen_at: string | null;
  resolved_at: string | null;
  evidence: RepositoryEvidence[];
};

export type RepositoryConfirmation = {
  kind: "fact" | "relationship";
  id: string;
  label: string;
  claim_status: string;
  human_resolution_status: "pending" | "confirmed" | "rejected";
  evidence: RepositoryEvidence[];
};

export type RepositoryDetailResponse = {
  workspace_id: string;
  mode: "repository_intelligence_read_only";
  source: "ri_006_persistence";
  repository: RepositoryIntelligenceIdentity;
  purpose: RepositoryFact | null;
  latest_audit: RepositoryLatestAudit | null;
  facts: RepositoryFact[];
  relationships: RepositoryRelationship[];
  findings: RepositoryFinding[];
  contradictions: RepositoryContradiction[];
  unknowns: RepositoryFact[];
  confirmation_queue: RepositoryConfirmation[];
  limitations: string[];
  truncated: {
    facts: boolean;
    relationships: boolean;
    findings: boolean;
    contradictions: boolean;
    confirmation_queue: boolean;
  };
  capabilities: RepositoryIntelligenceCapabilities;
};

export type RepositoryHistoryRun = {
  id: string;
  audit_level: "L0" | "L1" | "L2";
  target_status: "exact" | "unavailable";
  commit_sha: string | null;
  metadata_snapshot_id: string | null;
  profile: string;
  policy_hash: string;
  engine_version: string;
  status: "succeeded" | "partial" | "failed" | "cancelled";
  coverage_status: "complete" | "partial";
  completed_checks: string[];
  failed_checks: string[];
  skipped_checks: string[];
  limitations: string[];
  reconciliation_applied: boolean;
  artifact_count: number;
  artifact_status: "retained" | "purged";
  started_at: string;
  completed_at: string;
};

export type RepositoryHistoryResponse = {
  workspace_id: string;
  mode: "repository_intelligence_read_only";
  source: "ri_006_persistence";
  repository: RepositoryIntelligenceIdentity;
  runs: RepositoryHistoryRun[];
  limit: number;
  truncated: boolean;
  capabilities: RepositoryIntelligenceCapabilities;
};

export type RepositoryGraphNode = {
  id: string;
  full_name: string;
  repository_type: string;
  archived: boolean;
  open_findings_total: number;
  has_stale_intelligence: boolean;
  latest_audit_at: string | null;
};

export type RepositoryGraphEdge = {
  id: string;
  from_repository_id: string;
  from_repository_full_name: string;
  to_repository_id: string | null;
  target_full_name: string;
  relationship_type: string;
  resolution_status: "canonical" | "candidate";
  claim_status: "observed" | "inferred";
  human_resolution_status: "pending" | "confirmed" | "rejected";
  confidence: number;
  summary: string | null;
};

export type RepositoryGraphResponse = {
  workspace_id: string;
  mode: "repository_intelligence_read_only";
  source: "ri_006_persistence";
  nodes: RepositoryGraphNode[];
  edges: RepositoryGraphEdge[];
  summary: {
    nodes: number;
    edges: number;
    observed_edges: number;
    inferred_edges: number;
    candidate_edges: number;
  };
  truncated: {
    nodes: boolean;
    edges: boolean;
  };
  capabilities: RepositoryIntelligenceCapabilities;
};

export function buildRepositoryIntelligencePortfolioPath(
  workspaceId: string,
  limit = 100
): string {
  return `/api/v1/workspaces/${encodeURIComponent(
    workspaceId
  )}/repository-intelligence?limit=${encodeURIComponent(String(limit))}`;
}

export function buildRepositoryIntelligenceGraphPath(
  workspaceId: string,
  repositoryLimit = 200,
  edgeLimit = 500
): string {
  const params = new URLSearchParams({
    repository_limit: String(repositoryLimit),
    edge_limit: String(edgeLimit)
  });
  return `/api/v1/workspaces/${encodeURIComponent(
    workspaceId
  )}/repository-intelligence/graph?${params.toString()}`;
}

export function buildRepositoryIntelligenceDetailPath(
  workspaceId: string,
  repositoryId: string
): string {
  return `/api/v1/workspaces/${encodeURIComponent(
    workspaceId
  )}/repository-intelligence/repositories/${encodeURIComponent(repositoryId)}`;
}

export function buildRepositoryIntelligenceHistoryPath(
  workspaceId: string,
  repositoryId: string,
  limit = 20
): string {
  return `${buildRepositoryIntelligenceDetailPath(
    workspaceId,
    repositoryId
  )}/history?limit=${encodeURIComponent(String(limit))}`;
}

export async function fetchRepositoryIntelligencePortfolio(
  workspaceId: string,
  options: ApiFetchOptions = {}
): Promise<RepositoryPortfolioResponse> {
  return apiFetch<RepositoryPortfolioResponse>(
    buildRepositoryIntelligencePortfolioPath(workspaceId),
    options
  );
}

export async function fetchRepositoryIntelligenceGraph(
  workspaceId: string,
  options: ApiFetchOptions = {}
): Promise<RepositoryGraphResponse> {
  return apiFetch<RepositoryGraphResponse>(
    buildRepositoryIntelligenceGraphPath(workspaceId),
    options
  );
}

export async function fetchRepositoryIntelligenceDetail(
  workspaceId: string,
  repositoryId: string,
  options: ApiFetchOptions = {}
): Promise<RepositoryDetailResponse> {
  return apiFetch<RepositoryDetailResponse>(
    buildRepositoryIntelligenceDetailPath(workspaceId, repositoryId),
    options
  );
}

export async function fetchRepositoryIntelligenceHistory(
  workspaceId: string,
  repositoryId: string,
  options: ApiFetchOptions = {}
): Promise<RepositoryHistoryResponse> {
  return apiFetch<RepositoryHistoryResponse>(
    buildRepositoryIntelligenceHistoryPath(workspaceId, repositoryId),
    options
  );
}

export function normalizeRepositoryIntelligenceRepositoryId(
  value: string | null | undefined
): string | null {
  if (typeof value !== "string") {
    return null;
  }
  const normalized = value.trim().toLowerCase();
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/.test(
    normalized
  )
    ? normalized
    : null;
}
